# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/windows/TSFTextStore.h
# Commit: 1dc1117af560
# Full Hash: 1dc1117af56039488e28d5770e6379d1d89f4d19
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-02-08 17:32:27
# Regressor Bug: 1746104
# File Overlap Count: 1
# Description:
#   Bug 1746104 - part 1-4: Make `TSFTextStore` handle the no selection case r=m_kato
#   
#   Most change has already been done in part 1-2.  This patch does the remaining
#   work.
#   
# ==============================================================================

diff -r 1971c9f47541 -r 1dc1117af560 widget/windows/TSFTextStore.h
--- a/widget/windows/TSFTextStore.h	Mon Feb 07 22:33:31 2022 +0000
+++ b/widget/windows/TSFTextStore.h	Mon Feb 07 22:33:32 2022 +0000
@@ -19,6 +19,7 @@
 #include "mozilla/RefPtr.h"
 #include "mozilla/StaticPtr.h"
 #include "mozilla/TextEventDispatcher.h"
+#include "mozilla/TextEvents.h"
 #include "mozilla/TextRange.h"
 #include "mozilla/WindowsVersion.h"
 #include "mozilla/widget/IMEData.h"
@@ -567,8 +568,12 @@
       SetSelection(aSelectionChangeData);
     }
 
-    explicit Selection(uint32_t aStart, uint32_t aLength, bool aReversed,
-                       const WritingMode& aWritingMode) {
+    explicit Selection(const WidgetQueryContentEvent& aQuerySelectionEvent) {
+      SetSelection(aQuerySelectionEvent);
+    }
+
+    Selection(uint32_t aStart, uint32_t aLength, bool aReversed,
+              const WritingMode& aWritingMode) {
       SetSelection(aStart, aLength, aReversed, aWritingMode);
     }
 
@@ -602,6 +607,25 @@
                           aSelectionChangeData.GetWritingMode());
     }
 
+    bool SetSelection(const WidgetQueryContentEvent& aQuerySelectionEvent) {
+      MOZ_ASSERT(aQuerySelectionEvent.mMessage == eQuerySelectedText);
+      MOZ_ASSERT(aQuerySelectionEvent.Succeeded());
+      if (aQuerySelectionEvent.DidNotFindSelection()) {
+        if (mACP.isNothing()) {
+          return false;
+        }
+        mACP.reset();
+        // Let's keep the WritingMode because users don't want to change the UI
+        // of TIP temporarily since no selection case is created only by web
+        // apps, but they or TIP would restore selection at last point later.
+        return true;
+      }
+      return SetSelection(aQuerySelectionEvent.mReply->StartOffset(),
+                          aQuerySelectionEvent.mReply->DataLength(),
+                          aQuerySelectionEvent.mReply->mReversed,
+                          aQuerySelectionEvent.mReply->WritingModeRef());
+    }
+
     bool SetSelection(uint32_t aStart, uint32_t aLength, bool aReversed,
                       const WritingMode& aWritingMode) {
       const bool changed = mACP.isNothing() ||
